helper/fairshare/jobmanager.go

```diff
@@ -179,10 +177,94 @@ func (j *JobManager) getNextJob() Job {
 	}
 
 	if j.queues[queueID].Len() == 0 {
-		j.removeLastQueueAccessed()
+		// we remove the empty queue, but we don't remove the worker count
+		// in case we are still working on previous jobs from this queue.
+		// worker count cleanup is handled in j.decrementWorkerCount
+		delete(j.queues, queueID)
+	}
+
+	return out.(Job), queueID
+}
+
+// returns the next queue to assign work from, and a bool if there is a queue
+// that can have a worker assigned.
+// the intent is to avoid over-allocating work from specific queues, as
+// outlined in RFC VLT-145
+// note: this must be called with j.l held
+func (j *JobManager) getNextQueue() (string, bool) {
+	var nextQueue string
+	var canAssignWorker bool
+
+	queueIDsByIncreasingWorkers := j.sortByNumWorkers()
+	for _, queueID := range queueIDsByIncreasingWorkers {
+		if !j.queueWorkersSaturated(queueID) {
+			nextQueue = queueID
+			canAssignWorker = true
+			break
+		}
 	}
 
-	return out.(Job)
+	return nextQueue, canAssignWorker
+}
+
+// returns true if there are already too many workers on this queue
+// note: this must be called with j.l held (at least for read).
+// note: we may want to eventually factor in queue length relative to num queues
+func (j *JobManager) queueWorkersSaturated(queueID string) bool {
+	numActiveQueues := float64(len(j.queues))
+	numTotalWorkers := float64(j.workerPool.numWorkers)
+	maxWorkersPerQueue := math.Ceil(0.9 * numTotalWorkers / numActiveQueues)
```

Comment:
it might make sense to bump this to 0.95. It's 0.9 in the RFC, mainly to keep workers in reserve for new mounts (especially when the number of mounts is low). But 0.9 might keep too many workers in reserve. Hard to know without empirical data i think